### ISSUE TITLE
[problem] mmo_charstatus size & packets & random option issue.

### ISSUE THREAD

**Issue by anacondaqq** [rathena/rathena#1523]
Hello rAthena Developers.

My Issue related to this: https://rathena.org/board/topic/107087-compile-warnings-after-updating-to-hash-d20a7cafc0c32c0d6e2ad67e58f9ba45110d7ac7/
And to this: https://rathena.org/board/topic/107116-packet-size-increase-for-mmo-charstatus/
# Problem

A lot of us using not only classic rAthena src's, many of gm's using many custom src different features for their servers. And solution for minimizing some features just to have Random Option System -> bad solution.
# Question / Suggestion

Based on all information above, is it possible somehow safely take a look from developers to this problem, and somehow fix it? Under words "fix it", i mean : do some refactoring / variable optimizations. Or better -> increase somehow packet size to store for example not 65k bytes of data, but 128k, it will totally solve a issue. Please, don't suggest to me remove some custom features, or minimizing storage size, or anything else, it's bad idea to be clear.

**P.S.** i will be happy to help you to test anything.

**Comment by reunite-ro** [comment]
Hello i have some problem with this issue.

[Error]: mmo_charstatus size = 78664 is too big to be transmitted. (must be below 0xFFFF)

Max Storage 800. Thats only that ive changed

Can i know how to resolve this issue ?

**Comment by anacondaqq** [comment]
@reunite-ro look my comment here: https://github.com/rathena/rathena/pull/1524
it's because of random option issue and very big size of item structure (after update the size has been increased in around 1.8x times). 

The only one way (temporary fix) it's disable random option & remove from mmo_charstatus option_random. 

Or, much easier fix without disabling random option it's reducing in mmo.h next option:

> #define MAX_ITEM_RDM_OPT 5	 /// Max item random option [Napster]

from 5 to 1

> #define MAX_ITEM_RDM_OPT 1	 /// Max item random option [Napster]

it should reduce total size of item structure, and this change will free some memory in mmo_charstatus.

**Comment by mrjnumber1** [comment]
hmm, doing this will require a lot of changes, but it will likely be an issue down the road for large packets, like sending storage/gstorages ...

a lot of inter server packets use -1 for the len and use their 2nd word to tell packet len. allowing >65536 sized variable len packets will require changing to a long.. and unfortunately most of the packet handling code is adhesive towards exact byte lengths.. fixing it to allow more size would be quite annoying. 

having -1 and -2 packet len types for packets seems so hacky, eventually moving all would feel superfluous for most types..

i feel like the best option would be to do what should have been done long ago, and give all the packets their own struct so we'd have eg `struct packet_im_storage_fromsql *p; p->header = PHDR_IM_STORAGE_FROMSQL; p->len = .....` and just pass those around instead of these hideous beasts we have now.

**Comment by Everade** [comment]
mmo_charstatus is currently **extremly** limited.
 
## I was forced to:
- Remove extra storage
- Revert Max Party members back to official
- Revert Max Friends back to official

**YET I'M STILL getting disconnected!!!**

There's really NOT much i've increased.
Maybe guild members, and yet i'm still getting disconnected!!!!!!
 
In case this is not going to be improved real soon, they should remove ALL possibilities of those "now" useless settings.
Because it breaks the entire game. Majority of my players cannot login anymore.
 


This is a big piece of ********
# I would suggest and recommend to revert the implementation of these new features until the packet size problem has been solved.

If not, then ******* remove all the broken max settings!
Because there's about 0 freedom of customization left, unless you want to keep your player base from login in.

**Comment by secretdataz** [comment]
@Everade 

> PS:
I would be pleased to get github moderator rights so i can revert it myself.

You know you can revert commits on your local repo, right? :)

**Comment by anacondaqq** [comment]
@secretdataz this feature not well planned and not well implemented. 

If you think that only way where you can store structure is item structure, then, that means all other default game behavior will be very limited, that means people with any custom changes at their sources because of not available memory for operating with mmo_charstatus structure will always have errors.
In this case good suggestion will be next:
1. Create extra option in conf/feature.conf folder related to this random option feature.
2. Add define's for random option based on feature, and only then on client, for example:

```
#ifdef RANDOM_OPTION_FEATURE
... code of random option feature
#endif
```

These code should be always inside `#ifdef` to avoid any issues with the game client.

3. Then, to feature.conf need to write warning like: 

```
// Random Option Feature
// Minimal client 20150315
// Cause heavy item struct usage and a lot of problem with custom changes and modifications.
// By default disabled. To enable it go to mmo.h or core.h and uncomment #define RANDOM_OPTION_FEATURE
// Note: bla bla bla, this feature not yet well designed, use it on own risk.
// Parameter: 1-5 how many random options you can hold at the same time, by default 1
random_option_feature: 1
```

4. Turn off this feature FOR ALL by default in code, and write warning that this feature increase item struct to 1.8x  time, and other related to item structure in 2x times. And bla bla bla, warn admins to enable it on their own risk while complete fix will not be found.

5. Such method will solve maybe 99% of problems while someone will research and fix current architecture issues with item structure size.

**Comment by Everade** [comment]
@secretdataz
Of course i do, i'm not retarded.
The thing is that there are other good updates / map crash fixes or whatnot which we, as server owners want to implement.
I've also stated in my rage comment that people are being forced to hold updates back or create their own messed up workarounds like the one from @anacondaqq
(I appreciate your help though, but i'm not in the mood to go any further)

I'm just extremly pissed because we always have to fight with such ******** on our own, rather than being able to develop our own content instead.
Everything is possible if you have enough spare time to handle ********.
But i'm ******* done with this ****
Just by the time you manage to get something fixed yourself, the next ****storm is upon us.

Instead of pushing such updates on which most server owners have to struggle with, it should be kept back and being further improved until it's farely usable.
In short: Without affecting tons of servers content such as simple max settings.
But things are just getting worse.

So rather than wasting even more time on this, i will just delete 50k guilds out of my database and set guild members to 1.
That's probably gonna fix it.

**Comment by cydh** [comment]
yo bastard @Everade, don't raise ur voice to people that help you alive until now.
maybe ur last choice will be good, try delete all guilds in your server (is there any 50 guilds in your server?)

actually, I get nothing from help your issue personally, but try bump your head to keyword and hope this patch work https://gist.github.com/cydh/45f3c484a162dd1012216187880ec697

**Comment by Akkarinage** [comment]
There's really no need to be so damned rude, @Everade 

> I'm just extremly pissed because we always have to fight with such ******** on our own, rather than being able to develop our own content instead.

Then develop your own content. You don't have to rely on rA. Many RO servers have opted to add their code instead of wait for the rA dev team.

> Instead of pushing such updates on which most server owners have to struggle with, it should be kept back and being further improved until it's farely usable.

Nothing goes into the repo unless it's "farely usable". Or atleast, it works until you start customising the crap out of it. Sometimes things slip in that break the functionality of something else, but that's why we use a version control system. Major broken things can be reverted. Minor broken things can be fixed. Git is helpful. Git is life.

You can't expect to be able to customise something without repercussions, especially if you can't even be bothered to fix it yourself and then be at least a tiny bit helpful and throw the fix into a PR. That way, officially emulated mechanics aren't going to go completely out of the window if you change a few max settings.

> For sake WE DO, and we freaking NOTICE that some people cannot login anymore, but WE are the ****** people who have to find out what has been ******* ****** up yet again!

Where are your bug reports? Oh yeah:

> That's also one of the main reasons why i've stopped to report bugs.
Because if you don't give a ****, then i won't either.

You're right, the devs don't give a shit. They aren't giving up their free time to write code for you all to make money off from donations. Silly me, i thought this was a community project. Gives spoonfeeding a whole different meaning.

> So rather than wasting even more time on this, i will just delete 50k guilds out of my database and set guild members to 1.
That's probably gonna fix it.

Don't be some damned petulant. Atleast @anacondaqq is coming up with methods to combat the issue.

**Comment by Darkelfen** [comment]
By any chance did you defined random opts to 0 in mmo.h before throwing this monumental tantrum?
At least that worked for me using a gstor size of 999, and stor size of 600.

**Comment by aleos89** [comment]
You can check out #1115 which will help resolve this issue. I've just pushed some final fixes for it so the more testers we get with it the sooner we can push it to master!

**Comment by Jeybla** [comment]
Thanks @Darkelfen thats a pretty good workaround.

Your rant is inappropriate but I really understand you @Everade . I didn't merge rA since the random option system. Simply because of an increased storage size. And that's a very basic customization and nothing very fancy.

Even though I don't dislike the random option system patch. It's really important to merge big changes relatively quickly into the master to avoid later merge conflicts and mistakes by resolving these. That's why I'd like to see optimized inventory also merged into the master even if there are some minor issues. Maybe we should inform each other about higher issues in current releases via discord?

So back to topic: Why is the mmo_charstatus size limited and how can we increase it? Or is this not possible because of some client dependencies?

**Comment by Everade** [comment]
Sry guys, i had a really bad day.
But still, it's a bad situation for a lot of people, not just me.

My storage size= official
My extra storage = deleted because of this, _so official now_
Max Zeny = increased to 2'000'000'000
Max Friends = reduced back to official because of this
Max Party = reduced back to official because of this

### So in short:
The only thing i have increased right now is
max zeny
Maybe max HP is higher than usual

I can't think of anything else which should be increased.
And that's already causing the disconnection problem.
Which means the packet size, as it is, is either already hitting max, or is extremly close to it.


Hey luvely @cydh -12 years of data and top1 server for several years (the good old times you know). Do the math. There aren't as many players left, but the database is huge.
And thanks for the patch, going to take a look at it.

@Akkarinage I have reported bugs since eAthena. There's a reason i have the lv3 bug tracker award.
It's not like i've never helped out. I'm actually a nice person if things like these aren't ruining my day.

@aleos89 Thanks a lot! Will take my time to investigate it.

**Comment by reunite-ro** [comment]
I have this issue before and I changed this line in mmo.h

`#define MAX_ITEM_RDM_OPT 1 /// Max item random option [Napster]`

and no more issue :)

**Comment by aleos89** [comment]
Just wanted to note, keep in mind that if you double the packet size for the charstatus then you also double the network traffic required for each character that is connected. While it would solve this issue I don't think it's the best solution for this as the storage PR linked above should do wonders to resolve this issue.

**Comment by aleos89** [comment]
Fixed in ade1b17.

**Comment by Everade** [comment]
@aleos89 @lighta @cydh 

Thanks a lot!
That's a huge relief.

Really thank you for the optimisation/fix.
I appreciate it.
